### ISSUE TITLE
Streaming responses

### ISSUE THREAD

**Issue by janko-m** [socketry/falcon#7]
[tus-ruby-server](https://github.com/janko-m/tus-ruby-server) that I would like to run Falcon on can also be used to download uploaded files. In this case tus-ruby-server returns the response body as an object that responds to `#each` and lazily retrieves the file content from a storage (either disk, AWS S3 etc), which allows the web server to stream the content into the response body.

However, when I was looking at the Falcon source code, and I noticed that it wraps the response body that Rack apps return in `Async::HTTP::Body::Buffered`, which will eagerly retrieve all the chunks of a lazy response body object into memory, and only then continue writing it into the TCP socket.

That's unfortunately not ideal for my use case. First, these files can potentially be very large (i.e. 1 GB), so having the whole content loaded would cause high memory usage. Second, if the remote file is on AWS S3, the end user would have to wait for Falcon to download the whole file from AWS S3 before it can start downloading it.

It would be ideal if chunks of the response body would be written to the socket as they're retrieved, especially since Falcon can do it efficiently due to the async framework (unlike Puma which would block the web worker). May I ask what is the reason for Falcon buffering the response body?

**Comment by ioquatix** [comment]
I assume you are referring to https://github.com/socketry/falcon/blob/c1b762a4b52ea6bb2c9c1de2701cb882f7222c8a/lib/falcon/adapters/rack.rb#L100

I agree this is not ideal, I will need to review why what was added/changed, but I believe that the rack response was not sufficient for my purposes at the time. There is no reason why it needs to be fully buffered though. I will need to check exactly what is going on.

**Comment by ioquatix** [comment]
Okay, I did a little bit of digging.

`Buffered.wrap` is defined here:

https://github.com/socketry/async-http/blob/dd31bac812de504eb1c2b708b417627476f6ca8f/lib/async/http/body/buffered.rb#L29-L37

It does work efficiently if you give it an instance of `Body::Readable`. The other code paths also work, but the one that converts to an array is obviously less than ideal.

In supporting HTTP/1.x and HTTP/2.x, some issues became apparent to me. I am by no means an expert so feel free to correct me or advise me on something better.

There are really two kinds of responses (3 if you count `Connection: close`).

1. No length specified, send chunks of data. (`Transfer-Encoding: Chunked` or normal HTTP/2.0 data frames).
2. Length specified, send data. (`Content-Length` and write data, or normal HTTP/2.0 data frames).
3. No length. `Connection: Close`. Write data and close connection.

It's a bit more tricky because some HTTP/1.x semantics affect whether a body is sent or not. Like, a HEAD request will never expect a body. With HTTP/2.0, you can kind of ignore this, since the semantics of streams are not affected by the actual request/response payload.

That being said, it's not like there is a single perfect implementation. The `Async::HTTP::Body` implementation was my best effort to try and encapsulate streaming requests/responses. It mimics the HTTP/2.0 model which is essentially frames of data up to a given size. You can layer bodies (e.g. compression, decompression, etc). They work on chunks of data at a time, the chunk is determined by HTTP/2.0 flow control, or HTTP/1.0 socket buffer size or HTTP/1.1 chunk size.

Rack imposes it's own Input semantics which as I see you are fully aware, have shitty semantics w.r.t. memory usage and large input buffers.

Additionally, it gets complicated when you make requests. The problem is that when the request is idempotent, we should be able to retry. In that case, we need to re-send the body. I know this is different from Rack semantics, but I'm just trying to explain the motivation for the design. Right now, we don't retry requests if the request has a non-empty body.

https://github.com/socketry/async-http/blob/dd31bac812de504eb1c2b708b417627476f6ca8f/lib/async/http/request.rb#L34-L36

Coming back to the design of the protocol adapters, basically, I'm of the opinion that headers like `Transfer-Encoding` and `Content-Length` don't belong in the response from a Rack middleware. It's highly protocol specific. The body needs to capture that information.

Unfortunately rack encourages (e.g. `Rack::ContentLength`) protocol level decisions. Probably the only solution is to actually parse the response headers and body into an actual output, e.g. a custom `Falcon::Adapters::Output`. This can bind the accepted semantics of a rack response to an actual output body which can be understood by `async-http` protocol implementations.

**Comment by janko-m** [comment]
Thanks for the detailed explanation. However, I'm still not sure I understand why does the response need to be buffered. I was thinking that

* if the response headers include `Content-Length` (tus-ruby-server does this), we return a normal response with that `Content-Length`
* otherwise, if the size can be calculated from response body (which we can do if it responds to `#to_ary`, just like `Rack::ContentLength` does), calculate it, assign the result as `Content-Length`, and return a normal response
* otherwise return a `Transfer-Encoding: chunked` response

I think that's what typical web servers do (I checked Puma and it behaves like this).

Basically, what I'm asking is, what if in https://github.com/socketry/falcon/blob/c1b762a4b52ea6bb2c9c1de2701cb882f7222c8a/lib/falcon/adapters/rack.rb#L100

instead of `Async::HTTP::Body::Buffered` we used `Async::HTTP::Body::Fixed`? What would break?

**Comment by ioquatix** [comment]
- If the response headers include Content-Length (tus-ruby-server does this), we return a normal response with that Content-Length

Because trusting Content-Length can break HTTP/1.0 framing.

- Otherwise, if the size can be calculated from response body (which we can do if it responds to #to_ary, just like Rack::ContentLength does), calculate it, assign the result as Content-Length, and return a normal response

Yes, that's possible, and that's sort of what `Buffered` does already but in a slightly less efficient way.

- otherwise return a Transfer-Encoding: chunked response

Yes, feasible.

> instead of Async::HTTP::Body::Buffered we used Async::HTTP::Body::Fixed? Would anything break?

Yes, I suspect so.

**Comment by janko-m** [comment]
> Because trusting Content-Length can break HTTP/1.0 framing.

I think that it's up to the Rack application to ensure that the `Content-Length` it returned will match the actual number of bytes of the response body. For example, in tus-ruby-server I'm downloading a file from S3 and returning the content as an Enumerator as the response body. I already know the total bytesize of the S3 object upfront, so I also return the `Content-Length` response header, meaning the web server doesn't need to calculate the bytesize or use chunked response.

**Comment by ioquatix** [comment]
Breaking the protocol can lead to framing issues and security problems. In a proxy server, you might end up getting parts of a different request.

I understand where you are coming from, but I'm of the opinion `Content-Length` and `Transfer-Encoding` are protocol level headers and should never be specified by the application.

**Comment by ioquatix** [comment]
Incidentally that's why `Body#bytesize` exists, and in theory, it's still broken in HTTP1 because if you do something like `File.stat` to get file size, then start streaming it, you will have problems if the file changes part way through streaming.

**Comment by janko-m** [comment]
> I understand where you are coming from, but I'm of the opinion Content-Length and Transfer-Encoding are protocol level headers and should never be specified by the application.

After thinking about it, I think I understand what you mean.

Ok, so then according to Falcon philosophy the only way to stream content is using the `Transfer-Encoding: chunked` response format. I'm fine with that. Any objections in making that the behaviour when the Rack application returns a non-Array response body (instead of the current buffering approach)? If not, what would you suggest to for enabling streaming large response bodies?

I'm aware that another option would be to return a `Async::HTTP::Body::Fixed` or `Async::HTTP::Body::Writable` object for the response body, but I would prefer to make the tus-ruby-server agnostic to the web server.

**Comment by ioquatix** [comment]
I think chunked encoding is the only safe way but it sucks because it’s apparently mutually exclusive with content-length so the browser can’t show a progress bar? Or maybe I don’t understand the specs. Feel free to dig into this.

You can stream fixed size response if the body returns a valid bytesize.

You may run into some issues with mixing file io with async network operations.

**Comment by ioquatix** [comment]
(copied from chat so it doesn't get lost to the sands of time)

> Unrelated to what we discussed before, I thought I saw a month ago that Falcon had an ability to efficiently serve files on disk. Did I mix it up with some other project?

You can stream files efficiently using Falcon but it's non-trivial for a couple of reasons.
Firstly, streaming files using HTTP/2 requires encapsulation at the HTTP layer, you can't just use `sendfile` to pipe the IOs together, HTTP/2 simply doesn't allow for it. So, it's always going to be more CPU intensive.

I was thinking about this issue last night. It's probably not stupid to use HTTP/1.1 with `Connection: Close` and `Content-Length` for really large files.

Secondly, you have to consider how many streams of data you can actually read and send across the network simultaneously. This is non-trivial for a number of reasons - if you have more streams than physical disks, SSDs will be fine but spinning platters with high seek time and high throughput will be better if you have a worker pool 1-1 with the number of disks. It's tricky to make this work optimally.

Finally, there is no good async IO for disk files. Linux has something like `aio_read/aio_write` but I think they are just poor quality wrappers around `read/write` in glibc... something I should investigate further. The reality is however, disk IO is going to be much faster than network IO. Provided you use a relatively small buffer size, the latency of sequential reads should be minimal (since the OS will predict it). I suggest that you use a multiple of page sizes for a read buffer, and potentially investigate if you can use `madvise(SEQUENTIAL)` to reduce the latency. There is no actual issue using blocking IO in Async, just make sure you increase the number of running processes equal to the number of disks you have in your system (if you are IO bound) rather than the number of processors/cores (CPU bound).

Sounds like streaming files in the same way you stream any other content should be fine. I was recently even benchmarking this approach against sendfile, and I just didn't see any significant performance improvement (it was on localhost).

> Sounds like streaming files in the same way you stream any other content should be fine. I was recently even benchmarking this approach against `sendfile`, and I just didn't see any significant performance improvement (it was on localhost).

`sendfile` is a hack and it will be practically impossible to make it work well for http/2
As an addendum, if you are streaming the same file multiple times, if you read it from disk, it will end up in memory anyway, so it's worth considering that aspect of it too - eventually `read(fd)` will be very similar to memcpy

Even `memcpy` will "block" while it does it's job, but as far as I'm concerned, if you peg your CPU, you are doing fine. The problem with `puma` for example, is that if you do a lot of IO heavy work, you will max out your threads but not your CPU. Although, if you are reading from disk, in theory if you have a thread per disk, you will max out your IO subsystem, so perhaps that's okay.

**Comment by ioquatix** [comment]
I believe rack has an un-specified contract - if `body` responds to `to_path` it's representing a file on disk. It might be possible to add support for this and convert it to `Async::HTTP::Body::File`. There are still issues (it should really be tested/specced) when the file is truncated while being sent to the client. We should probably explicitly handle that case, either by zero fill, or terminating the response with an error.

**Comment by janko-m** [comment]
@ioquatix Thanks for the explanation! Ok, I think it's totally fine that Falcon doesn't have any special support for file streaming, I just wanted to check it. I now remembered that it was in [Iodine](https://github.com/boazsegev/iodine#static-file-serving-support) where I saw that feature.

> I believe rack has an un-specified contract - if body responds to to_path it's representing a file on disk. It might be possible to add support for this and convert it to Async::HTTP::Body::File. There are still issues (it should really be tested/specced) when the file is truncated while being sent to the client. We should probably explicitly handle that case, either by zero fill, or terminating the response with an error.

Yes, if response body object responds to `#to_path`, the `Rack::Sendfile` middleware will detect that and delegate the actual file serving to the frontend server (Nginx, Apache). And for me that's a perfectly viable alternative.

**Comment by ioquatix** [comment]
> Ok, I think it's totally fine that Falcon doesn't have any special support for file streaming

It does have `Async::HTTP::Body::File`.

> Yes, if response body object responds to #to_path

I will add support for this.

Another thing to think about is how `content-length` works when provided by Rack middleware. After gzip compression, the content-length will be different.

So, it's not trivial to figure out, e.g. for HEAD request.

**Comment by ioquatix** [comment]
Okay... while I'm not expecting this to change over night, here is a potential solution.

I've implemented `Falcon::Adapters::Input#clear` which removes data which has already been read.

It's possible to chop out the start of a file at the file system level, according to https://stackoverflow.com/questions/24239916/remove-beginning-of-file-without-rewriting-the-whole-file - So, `rack.input` that is based on a file on disk can "efficiently" implement this operation too.

In theory at least.

The next part of this, is that I've exposed `Falcon::Adapters::Input#body` which is the underlying streaming body. This could be exposed as part of the rack `env` and I'm fine with either approach.

Until we update the rack spec, this is the best we can do without either a/ violating the rack spec or b/ making the server horribly inefficient.

**Comment by ioquatix** [comment]
BTW, rewind still works after calling clear, but it only rewinds as far back as the last unread data.

**Comment by ioquatix** [comment]
Sorry, the above commit was referencing the wrong issue :)

**Comment by ioquatix** [comment]
Okay, I've implemented better support for the rack semantic model.

If the user gives content length, we will trust it to a certain extent, but it would be deleted from the headers and re-added as required by the protocol.

In a certain way of thinking, Falcon is acting like a proxy for the rack application. We should consider removing other `HOP_HEADERS`, e.g. `connection`, `keep-alive`, `transfer-encoding` and so on, because these will basically break the response generation (considering we handle transparently both HTTP/1 and HTTP/2).

I would love to have more specs for this area.
